criterion-measurement/src/Criterion/Measurement.hs

```diff
@@ -289,41 +291,69 @@ runBenchmark :: Benchmarkable
              -- exceeded in order to generate enough data to perform
              -- meaningful statistical analyses.
              -> IO (V.Vector Measured, Double)
-runBenchmark bm timeLimit = do
+runBenchmark bm timeLimit = runBenchmarkWith endAt 0 bm where
+  endAt :: Int -> Int64 -> Double -> NonEmpty Measured -> Double -> Maybe (Int64, Double)
+  endAt count iters delta ms prev = do
+    let m = NE.head ms
+
+    -- we accumulate the time spent in runs longer then threshold.
+    let prev' | measTime m >= threshold = prev + measTime m
+              | otherwise               = prev
+
+    -- We try to honour the time limit, but we also have more
+    -- important constraints:
+    --
+    -- We must generate enough data that bootstrapping won't
+    -- simply crash.
+    --
+    -- We need to generate enough measurements that have long
+    -- spans of execution to outweigh the (rather high) cost of
+    -- measurement.
+
+    if delta >= timeLimit && prev' >= threshold * 10 && count >= 4
+    then Nothing
+    else Just (nextIters iters, prev')
+
+  -- Multiply by 1.05, but always increase
+  nextIters :: Int64 -> Int64
+  nextIters n = max (n + 1) (div (105 * n) 100)
+
+-- | Run benchmark with provided dynamic termination condition.
+--
+-- The first argument is termination check, it takes as arguments:
+--
+-- * current run count (starting from zero)
+-- * current iteration count (starting from one)
+-- * amount of time the measurement process took so far.
+-- * results so far (in reverse order, latest first in the list).
+-- * user defined state
+--
+-- and should return 'Nothing' for run to terminate
+-- or @'Just' (nextIters, nextState)@ for run to continue with new iteration count and state.
```

Comment:
This needs a loud disclaimer that you should be careful when picking a termination check, as its implementation could affect the statistical quality of the data.

Reply:
[`analyseSample`](https://hackage.haskell.org/package/criterion-1.5.12.0/docs/Criterion-Analysis.html#v:analyseSample) doesn't have any pre condiotions specified. Looking at the code, it apperently needs at least 4 samples, then the statistical quality will be estimated (in the confidence interval).


I.e. I don't know how to put it "Termination check condition should be picked so the data generated is suitable for whatever further analysis is performed on it". Isn't that obvious?

Reply:
To be clear, I'm not looking for a super-detailed analysis of _how_ to make the data statistically significant. I'm just looking for a generic warning saying that you should exercise caution when using this function, as it has the potential to distort the meaningfulness of the data if implemented haphazardly.